Filament v3 Support (Fixes #23)

Log1x opened the pull request:
## Change log

- 🔧 Update `readOnly` to comply with Filament v3
- 🔧 Update the plugin `ServiceProvider` to support Filament v3
- 💄 Update Slug input style to match Filament v3
- 🔧 Update the `slug-input` field wrapper
- 🔧 Update Heroicons to v2
- 💄 Build the latest plugin assets
- 🚨 Run Pint

![Screenshot](https://i.imgur.com/RedTeON.png)
![Screenshot 2](https://i.imgur.com/1RN09Ll.png)

Log1x committed:
🔧 Update `readOnly` to comply with Filament v3

Log1x committed:
🔧 Update the plugin ServiceProvider to support Filament v3

Log1x committed:
💄 Update Slug input style for Filament v3
🔧 Update Heroicons to v2

Log1x committed:
⬆ Bump plugin requirements for Filament v3

Log1x committed:
💄 Build the latest plugin assets

Log1x committed:
🚨 Run Pint

Log1x commented:
```php
Camya\Filament\Forms\Components\TitleWithSlugInput::Camya\Filament\Forms\Components\{closure}(): Argument #2 ($set) must be of type Closure, Filament\Forms\Set given
```

Working on figuring out some changes with `$set` before this can be merged.

Log1x committed:
🔧 Update `$get` and `$set` types

Log1x commented:
Hmm this is very close to working. The exception above is fixed but when editing the slug it's not showing up once you press "OK" despite it properly storing it in the `hidden` field.

The slug is properly getting generated when creating an initial title, too.

howdu commented:
@Log1x I've pushed a couple of commits you can review. 

Adding `$wire.set` to manually push the changes fixed the issue you were having.

howdu committed:
chore: Update readme `slugRuleUniqueParameters` with `modifyRuleUsing`

fix: Fix manually changes slug
enhance: Use filament link component for external link

Log1x commented:
nice!! i appreciate it

camya commented:
Awesome, nice work @Log1x and @howdu. Thank you for the pull request. I just tested it a little bit.

**I found this problem:** 
If I type fast, the form skips some characters. It looks like the form is saved on each key stroke. (see flickering create button). I'm not shure what causes this. (See video below)

**Video:**

https://github.com/camya/filament-title-with-slug/assets/827658/46c72fc6-5f16-4a1f-94b2-4f03cd89ec30

Log1x commented:
I don't think it's actually saving, I think it's just swapping a network request to update the state due to `->reactive()` on the inputs.

It'd probably be a safe bet to change this to `->live(true)`.

**Edit**: I'm not seemingly able to reproduce the missed characters.
**Edit 2**: Nevermind, yes I am during Creation.

Log1x committed:
💚 Update test workflow to only use Laravel 10 and PHP 8.1

Log1x committed:
⬆ Bump dev-dependencies

Log1x committed:
♻️ Migrate the phpunit config

Log1x committed:
✅ Fix 2 of the failing tests

Log1x commented:
Updated the test stuff a bit but 3 of them are still failing due to https://github.com/livewire/livewire/blob/main/src/Features/SupportModels/ModelSynth.php#L20-L22

As far as I can tell the only fix for this on Livewire 3 is to do a database migration for the tests so we can create the Record using a RecordFactory. `->make()` and every other approach I could possibly think of does not work.

Log1x commented:
Ok I reproduced the character skipping when on a resource Create page. `->live(true)` instead of `->reactive()` is a "fix" but I think there's something else going on because it should be fine either way.

Log1x commented:
Hmm, I'm not sure. I never used the plugin in v2 so I don't know how it behaved, but commenting out the [`$set` for the field slug](https://github.com/camya/filament-title-with-slug/blob/43922293f0609c4bc2286ba5e8fafca772073911/src/Forms/Components/TitleWithSlugInput.php#L87-L89) fixed the issue as well as doing `->live(true)` instead to only update the slug on blur.

It's possible this is a weird quirk with Livewire 3 since it tries to bundle network requests. If we have 1 form getting live updated while also `$set`ting state on another component, I think it's overriding its self or something. I could be completely wrong though.

howdu commented:
I can't reproduce the flickering but I think the hidden input can be removed now it's manually being set from `$wire.set`

gafriputra commented:
can't wait for this PR to merge 🚀

faab007nl commented:
> can't wait for this PR to merge 🚀

Same here

Log1x committed:
🩹 Change input reactivity to blur

Log1x committed:
🔥 Remove redundant hidden input

Log1x commented:
I changed the input reactivity to use blur due to Livewire's network requests not being able to keep up with typing when on a "Create" page where it will auto-generate the slug for you. I think having it as blur still feels good and I'm having no issues on my end. 

Again, this issue only came into play on a Create page where the slug would get auto-generated based on the title you enter. Having it spam network requests to `Str::slug()` the slug in real-time while typing was otherwise causing it to not be able to keep up.

I also removed the hidden input that @howdu mentioned and everything seems good in my app.

I think it's possible for the `slug_auto_update_disabled` stuff to possibly get replaced as well but it's not a necessary change and everything works good as-is.

@camya do you want me to do a test database migration and fix the final 3 tests?

Log1x committed:
✅ Re-fix 2 of the failing tests

stvtk commented:
@camya @Log1x Do we know this PR is getting merged, please?

simplyaakif commented:
Any update on this. ? We appreciate you guys working on this.
We have a few resource that needs the v3 Update.

chosten commented:
A few suggestions:
- disableAutocomplete() is deprecated, use autocomplete(false) instead
- disableLabel() is deprecated, use hiddenLabel() instead
- use the `heroicon-o-arrow-top-right-on-square` icon instead of `heroicon-o-link` and add some padding between the text and the icon, or better maybe, use an [Action](https://filamentphp.com/docs/3.x/actions/trigger-button) instead so it will always match Filament style
- bonus: support field name using dot notation (eg: "slug.en") as described here: https://github.com/camya/filament-title-with-slug/pull/20#issuecomment-1456672758 but that's not related to Filament 3

Log1x commented:
If you want to do a quick PR against https://github.com/Log1x/filament-title-with-slug/tree/feat/filament-v3 I will merge it in – otherwise I don't mind doing these changes.

I also don't mind forking and maintaining this if @camya doesn't want to anymore. For now, I'm using it in my projects by adding the repository to my `composer.json`:

```json
{
  "require": {
    "camya/filament-title-with-slug": "dev-feat/filament-v3"
  },
  "repositories": [
    {
      "type": "vcs",
      "url": "https://github.com/log1x/filament-title-with-slug"
    }
  ]
}
```

chosten committed:
🗑️ Update deprecated methods

🩹 Fix visit link style in read only mode
🩹 Fix page reload when canceling or resetting the slug
💄 More consistent styling for the edit link\
💄 Add some space between links
✨ Support dot notation for attribute name

maikelenrise commented:
@Log1x I think it is pretty self evident that the op has lost interest in maintaining this package by now. I'd be very happy to see you fork and maintain this project!

chosten committed:
🩹 Make hidden field unique (#6)

chosten committed:
⚡️ Use `svg` directive instead of Blade component (#5)

Log1x committed:
⬆ Add support for Laravel 11

Log1x committed:
⬆ Bump dependencies for Laravel 12

mmoollllee commented:
Hey @Log1x can we get a Filamentv4 update? :) <3

howdu commented:
@mmoollllee https://github.com/broqit/filament-title-with-slug/tree/2.x
